vault/identity_store_aliases.go

```diff
@@ -256,8 +258,46 @@ func (i *IdentityStore) handleAliasUpdateCommon(req *logical.Request, d *framewo
 		return nil, err
 	}
 
+	var existingEntity *identity.Entity
+	if !newAlias {
+		// Verify that the combination of alias name and mount is not
+		// already tied to a different alias
+		if aliasByFactors != nil && aliasByFactors.ID != alias.ID {
+			return logical.ErrorResponse("combination of mount and alias name is already in use"), nil
+		}
+
+		// Fetch the entity to which the alias is tied to
+		existingEntity, err = i.MemDBEntityByAliasID(alias.ID, true)
+		if err != nil {
+			return nil, err
+		}
+
+		if existingEntity == nil {
+			return nil, fmt.Errorf("alias is not associated with an entity")
+		}
+		lockKeys = append(lockKeys, existingEntity.ID)
+	}
+
+	// Acquire the lock to modify the entity storage entry
+	locks := locksutil.LocksForKeys(i.entityLocks, lockKeys)
+	for _, lock := range locks {
+		lock.Lock()
+		defer lock.Unlock()
+	}
+
+	if canonicalID != "" {
+		entity, err = i.MemDBEntityByID(canonicalID, true)
+		if err != nil {
+			return nil, err
+		}
+		if entity == nil {
+			return logical.ErrorResponse("invalid canonical ID"), nil
+		}
+	}
```

Comment:
After this it should probably check again to see whether the Alias is still a part of this entity.